### ISSUE TITLE
Freeze stage1

### ISSUE THREAD

**Issue by thomie** [snowleopard/hadrian#250]
GHC developers use this feature daily.

https://ghc.haskell.org/trac/ghc/wiki/Building/Using#Freezingstage1

**Comment by snowleopard** [comment]
This can be done using Shake's `--skip` feature: https://github.com/ndmitchell/shake/issues/436. For example, one can pass `--skip=_build/stage1/*` to Hadrian to freeze Stage1.

@thomie Can you test how well this works?

**Comment by ndmitchell** [comment]
I wouldn't bother testing it just yet, wait for the next Shake.

**Comment by bgamari** [comment]
Andrey Mokhov <notifications@github.com> writes:

> This can be done using Shake's `--skip` feature:
> https://github.com/ndmitchell/shake/issues/436. For example, one can
> pass `--skip=_build/stage1/*` to Hadrian to freeze Stage1.
>
However, it would be great if the feature had a more memorable/less
awkward name. Having to know the build path of the stage 1 compiler
feels like a bit of a regression relative to the current build system.

**Comment by ndmitchell** [comment]
Once it is possible via `--skip`, adding sugar so it's a command line flag is easy enough.

**Comment by snowleopard** [comment]
@bgamari Sure, we can and will sugar it up nicely.

@ndmitchell Thanks for self-assigning!

**Comment by bgamari** [comment]
Andrey Mokhov <notifications@github.com> writes:

> @bgamari Sure, we can and will sugar it up nicely.
>
Awesome, thanks!

**Comment by hvr** [comment]
@bgamari fwiw, there's a several reasons why we'll keep around a legacy `Makefile`-wrapper anyway, so I think we'll just need to make sure that the legacy interface (c.f. #255) will translates the UI GHC developers are used to, to internal calls to `hadrian` with the respective flags/operations. I also think the legacy interface will remain the officially recommended way to interact with `hadrian`, and use `hadrian` directly only when requiring to access new features not part of the legacy UI.

**Comment by snowleopard** [comment]
> there's a several reasons why we'll keep around a legacy `Makefile`-wrapper anyway

@hvr What are these 'several reasons'? I'm aware of only one: don't break old scripts that call `make`.

>  I also think the legacy interface will remain the officially recommended way to interact with `hadrian`, and using `hadrian` directly is only desirable when requiring to access new features not part of the legacy UI.

But why? As soon as Hadrian is ready (hopefully this will happen some day), I'd recommend that everyone uses it directly to avoid any intermediate layers that introduce unnecessary noise.

**Comment by hvr** [comment]
> What are these 'several reasons'? I'm aware of only one: don't break old scripts that call make.

if you consider that only one reason (i.e. ci-scripts, local scripts, IDE integration, various tooling which detects `Makefile`s, etc), then I revise "several" into "a few" reasons :-)

The other reason I see is that I don't think that `hadrian` will be able to bootstrap itself. That's why we currently have a couple of awkward `build.sh` scripts, haven't we? So we need an additional layer anyway. This could actually be taken over by `boot`.

Another reason is that we're going to maintain `./boot` && `./configure` && `make` anyway. So we can just as well dogfeed it to make sure it doesn't break. I'm certainly going to keep using that workflow as there's no benefit to me, especially since `hadrian` seems to be developed with defaults which I consider wrong. So, personally, I won't recommend using `hadrian` directly either.

**Comment by snowleopard** [comment]
> So we need an additional layer anyway. This could actually be taken over by `boot`.

Oh, I can see how bootstrapping Hadrian via `make` can work well (at least on non-Windows machines), but please-please-please can we get rid of Perl scripts in GHC?! I was planning to subsume it into Hadrian some day...

The rest makes sense, thanks for providing more details. :+1:

**Comment by hvr** [comment]
I'm all for reducing reliance on `perl` in the long-term (I assume you're talking about `./boot`?) ... that would also finally allow us to drop bundling a `perl` interpreter for the GHC bindists for Windows (here I'm talking about https://ghc.haskell.org/ticket/9832).

**Comment by nomeata** [comment]
I just looked into using hadrian, but freezing stage1 seems to be a crucial feature for my usual workflow.

**Comment by snowleopard** [comment]
@nomeata It's probably possible to use the following interim solution while @ndmitchell is working on a proper one in Shake.

In the `Builder.hs` file we have:
```haskell
-- | Make sure a Builder exists on the given path and rebuild it if out of date.
needBuilder :: Builder -> Action ()
needBuilder = \case
    Configure dir -> need [dir -/- "configure"]
    builder       -> when (isInternal builder) $ do
        path <- builderPath builder
        need [path]
```
We could change this to:
```haskell
-- | Make sure a Builder exists on the given path and rebuild it if out of date.
needBuilder :: Builder -> Action ()
needBuilder = \case
    Configure dir -> need [dir -/- "configure"]
    Ghc _ Stage1  -> do
        path <- builderPath builder
        orderOnly [path]
    builder       -> when (isInternal builder) $ do
        path <- builderPath builder
        need [path]
```
This will schedule Stage1 GHC to be built before running it, but will not register it as a dependency.

Could you try if it works for you? Do a full clean rebuild just in case.

**Comment by nomeata** [comment]
> Could you try if it works for you? Do a full clean rebuild just in case.

Sorry, I have not actually employed hadrian yet, I was just exploring the option.

**Comment by bollu** [comment]
What is the status on this?

**Comment by snowleopard** [comment]
@bollu This is blocked by https://github.com/ndmitchell/shake/issues/436, and, ultimately, by https://github.com/ndmitchell/shake/issues/453.

Adding support for this feature in Hadrian will be quite simple once the above is complete.

**Comment by ndmitchell** [comment]
Yep, I'm still working on it, albeit on the back burner at the moment. Hopefully not too much longer.

**Comment by bgamari** [comment]
@ndmitchell what is the status of this? I believe this is one of the last remaining items before we can merge.

**Comment by ndmitchell** [comment]
@bgamari eek, I should really get a move on. I got a little distracted with other projects, but I'm now back on full focus. Hopefully not long.

**Comment by bgamari** [comment]
What is the status of this @ndmitchell? We would really like to merge Hadrian for GHC 8.4 but in order to do that we need to ensure that there is adequate time for developers, users, and packagers to use it and discover (and hopefully fix!) the bugs and usability issues that will inevitably come up. If the end of August arrives and we still haven't merged I fear that we may be pressed for time. Can we perhaps plan on finishing this up in the next couple of weeks? If not, what stop-gap solutions could we consider for freezing stage1?

**Comment by ndmitchell** [comment]
Sorry it's taken so long. Let's check back in a few weeks, but that's my goal.

**Comment by snowleopard** [comment]
@bgamari If the Shake `--skip` feature won't be ready in time, I think we can hack up something in Hadrian, for example using `orderOnly` dependencies on stage1 build artefacts instead of `need` when we are in the freezeStage1 mode. I mentioned this here: https://github.com/snowleopard/hadrian/issues/250#issuecomment-249750913

**Comment by ndmitchell** [comment]
It's slowly bubbling to the top of one of my priority lists. The fact that I'm going through tickets is a good sign at least!

**Comment by ndmitchell** [comment]
I now have Hadrian building with shake-HEAD. I think the release should be imminent - just one last bug/change to go.

**Comment by izgzhen** [comment]
> I now have Hadrian building with shake-HEAD. I think the release should be imminent - just one last bug/change to go.

How does everything go so far? Since I've basically fixed the cross compilation issue, so I can also help on this one if plausible.

**Comment by bgamari** [comment]
@ndmitchell, what is the status of this?

**Comment by ndmitchell** [comment]
I just need a free evening to cut the release - basically done.

**Comment by snowleopard** [comment]
I've added a draft implementation for this feature -- see the above commit. To use it, pass `--freeze1` when running Hadrian. Seems to work well in my experiments, but it would be nice to hear feedback from others.

I'll also need to add docs and see if I can simplify code.

**Comment by izgzhen** [comment]
I've tried it a bit and it looks working on my Linux machine, good job!

**Comment by snowleopard** [comment]
Thanks @izgzhen!

And thanks to @ndmitchell for making the implementation so straightforward :)

**Comment by snowleopard** [comment]
OK, I think we can close this issue. Please shout if I missed anything.
